# Graded Challenge 1 (DA1 and DA2)

Now that we've seen how Python works and and how to use Pandas, it's time for you to combine and apply this knowledge to analyse social media data.

The warm-up challenge consists of two components:
1. **Programming**
2. **Interpretation**

**Some important notes for the challenges:**
1. While we of course like when you get all the answers right, the important thing is to exercise and apply the knowledge. So we will give some credit for challenges that may not be complete, as long as we see enough effort. The rubric (see Canvas) reflects this.
2. You will deliver the challenges on Canvas. Make sure to follow the turning-in instructions. 

### Usage of AI-generated content

In line with the course policies, it is **not allowed** to use AI-generated content to any explanations or text provided in MarkDown. This means that for **Interpretation** students are **not allowed** to use any AI-generated content. 

For **Programming**, however, AI-assisted tools may be helpful to understand how to solve some of the questions - as one may also use Google to search for solutions online (e.g., on StackOverflow or other websites). They can, therefore, be used for the programming part of this challenge exceptionally - and only for this part - unless otherwise noted in the question. It is however your responsibility to test and make sure that the solution works - and explain this, in your own words, in the interpretation section.

### Facing issues? 

We are constantly monitoring the issues on the GitHub to help you out. Don't hesitate to log an issue there, explaining well what the problem is, showing the code you are using, and the error message you may be receiving. 

**Important:** We are only monitoring the repository in weekdays, from 9.30 to 17.00. Issues logged after this time will most likely be answered the next day. This means you should not wait for our response before submitting a challenge :-)

## 1. Programming

Select one of the datasets available for this class. Load that file using ```pandas``` and answer the following questions:
1. How many cases are in your dataset?
2. What columns does the dataset contain and what is their data type?
3. Are there any missing values?
4. What is the lowest number, highest number, average and the standard deviation of engagement (you can choose your metric)?
5. What is the most common language of the entries?
6. Who are the 5 most frequent authors in your dataset (can be poster, advertiser or subreddit)?
7. Write a function that categorizes the content of your dataset based on information available in one of the columns (e.g., engagement, some characteristics,  or any other column). It's up to you to define what this function should do (and explain it). One possibility is to create a function that returns 1 if the row belongs to a certain category, and 0 if it does not. You can be more creative than this (see Rubric on Canvas). *This week, we ask you to write the function and test if it works (following the tutorial). You will learn how to apply it to the dataframe next week.*

**Note on question 7.** We do not expect you to use  ```text``` column for the function. We will learn how to categorize that column in the next tutorials. For question 7 you are **not** allowed to use AI-assisted tools.

### Using MarkDown

Make sure to combine code and markdown to answer these questions. Mention specifically the question (and question number) and the answer in markdown, relating to the code and the output of the code. Failing to do will impact the grade, as we will not be able to see whether you answered the question.

*Wordcount: There is no maximum wordcount for this section.*


In [1]:
import pandas as pd
import json


pad_naar_bestand = "/Users/jensvanasperen/GradedChallenge1/mcdonalds_or_@mcDonalds.jsonl" 

all_tweets = []
with open(pad_naar_bestand, 'r') as f:
    for line in f:
        line_data = json.loads(line)
        if 'data' in line_data:
            all_tweets.extend(line_data['data'])

df = pd.json_normalize(all_tweets)
df.head()

created_at      conversation_id lang  possibly_sensitive  \
0  2023-02-07T15:29:33.000Z  1622980893416517632   en               False   
1  2023-02-07T15:29:08.000Z  1622980790761013255   en               False   
2  2023-02-07T15:29:08.000Z  1622980787946635264   de               False   
3  2023-02-07T15:29:07.000Z  1622980784234840064   en               False   
4  2023-02-07T15:29:05.000Z  1622980778622857216   en               False   

             author_id reply_settings edit_history_tweet_ids  \
0  1547500137026551809       everyone  [1622980893416517632]   
1  1015653575916126211       everyone  [1622980790761013255]   
2  1356299130994388994       everyone  [1622980787946635264]   
3           2989505299       everyone  [1622980784234840064]   
4   859693365058617344       everyone  [1622980778622857216]   

                                   referenced_tweets                   id  \
0  [{'type': 'retweeted', 'id': '1622902416868884...  1622980893416517632   
1                                                NaN  1622980790761013255   
2                                                NaN  1622980787946635264   
3  [{'type': 'retweeted', 'id': '1622721798902251...  1622980784234840064   
4  [{'type': 'retweeted', 'id': '1622721798902251...  1622980778622857216   

                                                text  ...  \
0  RT @SaulStaniforth: "I'm paid about £13 an hou...  ...   
1  Trade: Brian Windhorst is sending $11.36 to a ...  ...   
2  Falschparker Thread ist neue McDonalds Thread....  ...   
3  RT @weirdlilguys: u got mcdonalds money? https...  ...   
4  RT @weirdlilguys: u got mcdonalds money? https...  ...   

                                entities.annotations  \
0                                                NaN   
1  [{'start': 7, 'end': 21, 'probability': 0.8675...   
2                                                NaN   
3  [{'start': 24, 'end': 32, 'probability': 0.612...   
4  [{'start': 24, 'end': 32, 'probability': 0.612...   

                                       entities.urls   attachments.media_keys  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3  [{'start': 41, 'end': 64, 'url': 'https://t.co...  [3_1622464177188093952]   
4  [{'start': 41, 'end': 64, 'url': 'https://t.co...  [3_1622464177188093952]   

   in_reply_to_user_id  entities.cashtags  entities.hashtags  geo.place_id  \
0                  NaN                NaN                NaN           NaN   
1                  NaN                NaN                NaN           NaN   
2                  NaN                NaN                NaN           NaN   
3                  NaN                NaN                NaN           NaN   
4                  NaN                NaN                NaN           NaN   

  attachments.poll_ids geo.coordinates.type geo.coordinates.coordinates  
0                  NaN                  NaN                         NaN  
1                  NaN                  NaN                         NaN  
2                  NaN                  NaN                         NaN  
3                  NaN                  NaN                         NaN  
4                  NaN                  NaN                         NaN  

[5 rows x 30 columns]

#1. How many cases are in your dataset?


In [2]:
print(f"1. Total number of cases: {len(df)}")

1. Total number of cases: 12331


##To determine the dataset size, I used the len(df) function. This shows that the dataset contains 12,331 individual cases (tweets).

#2. What columns does the dataset contain and what is their data type?

In [3]:
print("Overview of data types:")
print(df.dtypes)

Overview of data types:
created_at                            str
conversation_id                       str
lang                                  str
possibly_sensitive                   bool
author_id                             str
reply_settings                        str
edit_history_tweet_ids             object
referenced_tweets                  object
id                                    str
text                                  str
public_metrics.retweet_count        int64
public_metrics.reply_count          int64
public_metrics.like_count           int64
public_metrics.quote_count          int64
public_metrics.impression_count     int64
edit_controls.edits_remaining       int64
edit_controls.is_edit_eligible       bool
edit_controls.editable_until          str
entities.mentions                  object
context_annotations                object
entities.annotations               object
entities.urls                      object
attachments.media_keys             object
in_reply_t

##Inspected the data types using df.dtypes. The output shows a mix of different types. For example 'object' or 'bool'.

#3. Are there any missing values?

In [4]:
print("Missing values per column:")
print(df.isna().sum())

Missing values per column:
created_at                             0
conversation_id                        0
lang                                   0
possibly_sensitive                     0
author_id                              0
reply_settings                         0
edit_history_tweet_ids                 0
referenced_tweets                   1527
id                                     0
text                                   0
public_metrics.retweet_count           0
public_metrics.reply_count             0
public_metrics.like_count              0
public_metrics.quote_count             0
public_metrics.impression_count        0
edit_controls.edits_remaining          0
edit_controls.is_edit_eligible         0
edit_controls.editable_until           0
entities.mentions                   1386
context_annotations                 1120
entities.annotations                5832
entities.urls                       5726
attachments.media_keys              6674
in_reply_to_user_id           

In [5]:
total_missing = df.isna().sum().sum()
print(f"\nTotal missing values in dataset: {total_missing}")


Total missing values in dataset: 104860


##Checked for missing values using isna().sum(). The output reveals that core columns like 'text' have 0 missing values, while optional fields like 'geo.place_id' have many.This confirms the dataset is complete enough for the required analysis.

#4. What is the lowest number, highest number, average and the standard deviation of engagement (you can choose your metric)?

In [6]:
print("\n4. Engagement Statistics (Likes):")
stats = df['public_metrics.like_count'].describe()
print(stats)


4. Engagement Statistics (Likes):
count    12331.000000
mean         0.858892
std         13.808035
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       1160.000000
Name: public_metrics.like_count, dtype: float64


##Calculated the descriptive statistics for engagement.The output shows a mean of 0.86 and a maximum of 1,160 likes. This indicates that while most tweets have very low engagement, there are a few significant outliers in the dataset.

#5. What is the most common language of the entries?

In [7]:
common_lang = df['lang'].mode()[0]
print(f"\n5. Most common language: {common_lang}")


5. Most common language: en


##Identify the primary audience. The output is 'en', meaning English is the dominant language in this McDonald's dataset.

#6. Who are the 5 most frequent authors in your dataset (can be poster, advertiser or subreddit)?

In [8]:
print("\n6. Top 5 most frequent authors (IDs):")
print(df['author_id'].value_counts().head(5))


6. Top 5 most frequent authors (IDs):
author_id
60421964               110
71026122                70
1470764537418862594     30
2388019676              22
1343141573622718466     19
Name: count, dtype: int64


##Identify the top 5 most active authors by their ID. The output lists the top 5 Author IDs, with the most frequent author appearing 110 times. This identifies the most prominent voices in the conversation.

#7. Write a function that categorizes the content of your dataset based on information available in one of the columns.

In [9]:
def categorize_engagement(likes):
    """
    This function categorizes a tweet based on the number of likes.
    1 = High engagement (more than 10 likes)
    0 = Standard engagement (10 likes or less)
    """
    if likes > 10:
        return 1
    else:
        return 0

In [11]:
# Apply the function to the 'public_metrics.like_count' column 
# and create a new column called 'engagement_labeldf['engagement_label'] = df['public_metrics.like_count'].apply(categorize_engagement)
df['engagement_label'] = df['public_metrics.like_count'].apply(categorize_engagement)

In [12]:
# Show the first 10 rows to see the new column
print("Result of the categorization:")
print(df[['public_metrics.like_count', 'engagement_label']].head(10))

Result of the categorization:
   public_metrics.like_count  engagement_label
0                          0                 0
1                          1                 0
2                          1                 0
3                          0                 0
4                          0                 0
5                          0                 0
6                          0                 0
7                          0                 0
8                          0                 0
9                          0                 0


In [13]:
# Count how many tweets fall into each category
print("\nValue counts of the new categories:")
print(df['engagement_label'].value_counts())


Value counts of the new categories:
engagement_label
0    12187
1      144
Name: count, dtype: int64


##Wrote a custom Python function to categorize tweets into 'High Engagement' (1) or 'Standard' (0) based on a threshold of 10 likes. There are 12187 'Standard' tweets and 144 'High Engagement' tweets.

## 2. Intepretation

### Explaining your process

Write a brief description explaining how you created or assembled the code (i.e., sources used, and how did you adjust the code). In this description, include a brief reflection of the main challenges you had during the process, and how you handled these challenges.

*Wordcount: maximum of 200 words. Please inform the wordcount for this answer*

To assemble the code, I followed the CRISP-DM framework focusing on the Data Understanding and Preparation phases. I used the Pandas library to handle the McDonald's dataset, specifically using pd.json_normalize to flatten the nested JSONL structure into a readable table. This allowed me to access columns like public_metrics.like_count for statistical analysis.

The main challenge was the complex structure of the JSONL data, where essential metrics like "like_count" were nested within a dictionary called "public_metrics". Initially, Pandas could not perform calculations because it saw the entire dictionary as a single object. I handled this by applying a normalization technique to get the data into individual columns. Another challenge was verifying data quality; I had to learn that Pandas labels text as 'object' for example. Finally, I adjusted the code for question 7 by creating a custom 'if-else' function to categorize engagement, ensuring the data was correctly prepared for interpretation.


Wordcount: 148

### Explaining your code

You used a lot of code above to answer the seven questions. Select what you consider to be the three most important steps in the code, justify why they are important, and explain the commands being used. 

*Note: there are different correct answers to what the three most important steps of the code are. We want to understand your reasoning/justification for this, and will accept different answers depending on the logic and clarity of your argumentation*.

*Wordcount: maximum of 200 words. Please inform the wordcount for this answer*

Data Normalization: This was the first essential step in the process. Since the social media data was structured in complex layers, the information was flattened into a standard table format. This process is crucial because it makes hidden information, such as engagement metrics, accessible for analysis.

Descriptive Statistical Analysis: This step provided a simple view of the dataset during the exploration phase. By calculating values like the mean and maximum, it was possible to understand the typical reach of a tweet and identify significant outliers in the McDonald's data.

Applying a Categorization Logic: A specific method was used to transform raw numbers into meaningful categories. This is a vital part of preparing the data, as it allows for a clear distinction between standard interactions and high-impact content. This makes the final insights much more actionable.

Wordcount: 135

### Quality assurance

You are assembling code from different sources - some of it you learned in the tutorial, you may have reused code we provided, and in some cases you may have used different online sources. But how do you know that the code actually worked and delivered the expected results? Please explain how someone else can verify that the code worked (e.g., what should they look at) and which steps you built into the code for yourself to know that it worked (explicitly indicate at least three steps)

*Wordcount: maximum of 200 words. Please inform the wordcount for this answer*

To guarantee that the code functioned correctly and delivered the expected results, several validation steps were integrated throughout the process. An external reviewer can verify the accuracy of the findings by inspecting the printed outputs directly beneath each code cell, which provide visual confirmation of the data transformations.

Specifically, three primary verification mechanisms were built into the workflow to monitor performance. First, a row count verification was performed immediately after loading the dataset to confirm that the entire file was imported without data loss. Second, quality checks were executed on data types and missing values to ensure that all calculations were based on clean, numerically correct data. Finally, a visual sampling was conducted by displaying the first five rows of the modified table and generating frequency counts for the new categories. This allowed for a manual check to ensure that the categorization logic was applied as intended to the final dataset. These combined measures ensure that the final output is reliable.

Wordcount: 161 